File: _scripts/getdata.py
#!/usr/bin/env python

import requests
import json
from pprint import pprint
import math
import sys

def calculate_distance(lat1, lon1, lat2, lon2):
	# return math.sqrt(math.pow(user_lat - feature_lat, 2) + math.pow(user_lon - feature_lon, 2))

	R = 6371 # radius of the earth in km
	dLat = deg_to_rad(lat2 - lat1)
	dLon = deg_to_rad(lon2 - lon1)
	a = (
	     math.sin(dLat/2) * math.sin(dLat/2) +
	     math.cos(deg_to_rad(lat1)) * math.cos(deg_to_rad(lat2)) *
	     math.sin(dLon/2) * math.sin(dLon/2)
	     )
	c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
	d = R * c # distance in km
	return d

def deg_to_rad(deg):
	return deg * (math.pi / 180.0) 


# handle lat/lon args
has_user_location = False
if len(sys.argv) == 3:
	has_user_location = True
	user_lat = float(sys.argv[1])
	user_lon = float(sys.argv[2])

# fetch data
r = requests.get('https://anrmaps.vermont.gov/arcgis/rest/services/Open_Data/OPENDATA_ANR_TOURISM_SP_NOCACHE_v2/MapServer/163/query?where=1%3D1&outFields=*&outSR=4326&f=json')
data = r.json()['features']


# manipulate data
for entry in data:
	# add distnace to user
	if has_user_location:
		feature_lat = float(entry['geometry'].get('y'))
		feature_lon = float(entry['geometry'].get('x'))
		dist_km = calculate_distance(feature_lat, feature_lon, user_lat, user_lon)
		dist_deg = math.sqrt(math.pow(user_lat - feature_lat, 2) + math.pow(user_lon - feature_lon, 2)) # distance formula
		# print("dist:", dist)
		entry.update(distance={'km': dist_km, 'deg': dist_deg})

	# sanitize 'No' and 'Yes' values
	for key, value in entry['attributes'].items():
		# print(key, "=", value)
		if value == 'No':
			entry['attributes'][key]=False
		elif value == 'Yes':
			entry['attributes'][key]=True


# pprint(data)

# return json string to php
json_string = json.dumps(data)
print(json_string)

File: _scripts/getdata.php
<?php
/**
 * Created by PhpStorm.
 * User: zach
 * Date: 12/20/17
 * Time: 3:46 PM
 */

/* ### fetch data (python) [OBSOLETE] ###
$cmd = escapeshellcmd("python $root/_scripts/getdata.py 44.518087699999995 -73.18415689999999");
$output = shell_exec($cmd);
$data = json_decode($output, true);
*/


/** calculates great-circle distance between two points on sphere
 *
 * reference: https://en.wikipedia.org/wiki/Haversine_formula
 * @param $lat1 float|int pos1 latitude
 * @param $lon1 float|int pos1 longitude
 * @param $lat2 float|int pos2 latitude
 * @param $lon2 float|int pos2 longitude
 * @return float|int the distance in km between the two points
 */
function calculate_distance($lat1, $lon1, $lat2, $lon2) {
    $earth_radius = 6371.0; # radius of the earth in km
    $dLat = deg2rad($lat2 - $lat1);
    $dLon = deg2rad($lon2 - $lon1);

    $hav    = sin($dLat/2) * sin($dLat/2) + cos(deg2rad($lat1)) * cos(deg2rad($lat2)) * sin($dLon/2) * sin($dLon/2); // haversine of central angle
    $havInv = 2 * asin(sqrt($hav)); // inverse haversine function
    $dist   = $earth_radius * $havInv; # distance in km

    return $dist;
}


#  user location - todo: make this dynamic
// currently hardcoded to green mountain power in colchester, vt
$hasUserLocationData = true;
$userLat = 44.518087699999995;
$userLon = -73.18415689999999;


# get data
$jsonStr = file_get_contents("https://anrmaps.vermont.gov/arcgis/rest/services/Open_Data/OPENDATA_ANR_TOURISM_SP_NOCACHE_v2/MapServer/163/query?where=1%3D1&outFields=*&outSR=4326&f=json")
        ?: file_get_contents("$root/_tmp/json_backup.json"); // file_get_contents() returns false if it fails -- in that case, use a local backup.
$data    = json_decode($jsonStr, true)["features"];


# mutate data
foreach ($data as $idx => $entry) {
    # add distance from user
    if ($hasUserLocationData) {
        $featureLat = $entry["geometry"]["y"];
        $featureLon = $entry["geometry"]["x"];
        $distKm  = calculate_distance($featureLat, $featureLon, $userLat, $userLon);
        $distDeg = sqrt(pow($userLat - $featureLat, 2) + pow($userLon - $featureLon, 2)); # distance formula

        $data[$idx] += array("distance" => array("km" => $distKm, "deg" => $distDeg));
    }

    # sanitize yes, no, and n/a values
    foreach ($entry["attributes"] as $attr => $value) {
        switch (strtolower($value)) {
            case "no"  : $data[$idx]["attributes"][$attr] = false; break;
            case "yes" : $data[$idx]["attributes"][$attr] = true;  break;
            case "n/a" : $data[$idx]["attributes"][$attr] = null;  break;
        }
    }
}

// print data array
if ($debug) {
    print "<pre>";
    print '$data array:'. PHP_EOL;
    print_r($data);
    print "</pre>";
}